### PR TITLE
[Merged by Bors] - feat: Pontryagin duality for finite abelian groups

### PR DESCRIPTION
**Pull request by YaelDillies**
Prove Pontryagin duality in case of finite abelian groups.

This states that any finite abelian group is canonically isomorphic to its double dual (the space of complex-valued characters of its space of complex-valued characters).

We first prove it for `ZMod n` and then extend to all finite abelian groups using the Structure Theorem.

From LeanAPAP

Co-authored-by Bhavik Mehta <bhavikmehta8@gmail.com>

---

- [x] depends on: #15441
- [x] depends on: #16447

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/SpecialFunctions/Complex/CircleAddChar.lean
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Complex/CircleAddChar.lean:26]: We should relate this def to the ones below, eg toCircle. I think that one is a composition of this and the (not really) canonical map from ZMod N to R given by x -> x/N

File: Mathlib/Analysis/SpecialFunctions/Complex/PontryaginDuality.lean
Comment (b-mehta) [inline Mathlib/Analysis/SpecialFunctions/Complex/PontryaginDuality.lean:64]: Is it any better to write this in terms of `toCircle`? It's a group hom in ZMod n composed with that; maybe this also saves some of the lemmas above

Comment (YaelDillies) [reply Mathlib/Analysis/SpecialFunctions/Complex/PontryaginDuality.lean]: It is better!

File: Mathlib/Analysis/Fourier/FiniteAbelian/Orthogonality.lean
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/Fourier/FiniteAbelian/Orthogonality.lean:37]:
```suggestion
  simp only [Fintype.expect_eq_sum_div_card, AddChar.sum_eq_ite ψ, h, ↓reduceIte, zero_div]
```
There is already some API for `AddChar`s...

File: Mathlib/Analysis/Fourier/FiniteAbelian/Orthogonality.lean
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/Fourier/FiniteAbelian/Orthogonality.lean:80]:
```suggestion
  exact linearIndependent_of_ne_zero_of_wInner_cWeight_eq_zero AddChar.coe_ne_zero
    fun ψ₁ ψ₂ ↦ wInner_cWeight_eq_zero_iff_ne.2
```

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:54]:
```suggestion
  intro x y h
```

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:55]:
```suggestion
  have hn : (n : ℝ) ≠ 0 := mod_cast NeZero.ne n
```
`replace` does not make sense here; there is no previous `hn` in context.

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:97]: With the additional imports
```
import Mathlib.GroupTheory.FiniteAbelian.Duality
import Mathlib.RingTheory.RootsOfUnity.AlgebraicallyClosed
import Mathlib.Analysis.Complex.Polynomial.Basic
```
one can use the results on homomorphisms from finite abelian multiplicative groups into monoids with enough roots of unity, as follows.
```
variable (α) in
/-- The additive isomorphism from `AddChar α M` to homomorphisms from `Multiplicative α` to `M`. -/
def mulEquivMultiplicativeMonoidHom (M) [CommMonoid M] :
    Multiplicative (AddChar α M) ≃* (Multiplicative α →* M) :=
  let e : Multiplicative (AddChar α M) ≃* AddChar α M :=
    MulEquiv.mk' Additive.ofMul fun _ _ ↦ rfl
  e.trans <| AddChar.toMonoidHomMulEquiv

variable (α) in
/-- The group of complex-valued additive characters on a finite abelian group is
(noncanonically) isomorphic to itself. -/
lemma addEquiv_self [Finite α] : Nonempty (AddChar α ℂ ≃+ α) :=
  let e := (CommGroup.mulEquiv_monoidHom_of_hasEnoughRootsOfUnity (Multiplicative α) ℂ).some
  let e₀ : (Multiplicative α →* ℂ) ≃* (Multiplicative α →* ℂˣ) := {
    toFun := MonoidHom.toHomUnits
    invFun := fun φ ↦ (Units.coeHom _).comp φ
    left_inv := fun φ ↦ by ext; simp
    right_inv := fun φ ↦ by ext; simp
    map_mul' := fun φ ψ ↦ by ext; simp
  }
  let e' := (mulEquivMultiplicativeMonoidHom α ℂ).trans <| e₀.trans e.symm
  ⟨AddEquiv.toMultiplicative.symm e'⟩

@[simp] lemma card_eq [Fintype α] : card (AddChar α ℂ) = card α :=
  Fintype.card_congr (addEquiv_self α).some.toEquiv
```

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:104]: One could then golf this to
```lean
def zmodAddEquiv : ZMod n ≃+ AddChar (ZMod n) ℂ :=
  addEquiv_self (ZMod n) |>.some.symm
```
but I assume you what the specific isomorphism here.

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (MichaelStollBayreuth) [inline Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:126]: It should be possible to reduce this to `CommGroup.exists_apply_ne_one_of_hasEnoughRootsOfUnity (Multiplicative α) ℂ`, but it looks like this requires building some API for `mulEquivMultiplicativeMonoidHom ` from above (and possibly variants).

So, on further thought, I think it is not worth the trouble, and it is OK to leave the code basically as is.

**Review by MichaelStollBayreuth — APPROVED** [review]
LGTM, modulo one golf and a couple minor style issues.

**Comment by MichaelStollBayreuth** [comment]
maintainer delegate

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (b-mehta) [reply Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:104]: Yes, the actual isomorphism here was definitely used in the original application of this code about a year and a half ago; I'm pretty sure it's useful in the application in LeanAPAP too.

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (b-mehta) [reply Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:97]: Do you think there should be a TODO in this file so it doesn't get forgotten? Or do you think the linkage here isn't worth it in general

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (YaelDillies) [reply Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:54]: I don't really see the point of learning when to use `intro` when it is strictly weaker than `rintro`. If you ask me, I would be very happy for mathlib to make `intro` notation for `rintro`

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (YaelDillies) [reply Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:104]: I actually don't think we ever use the specific isomorphism, no

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (YaelDillies) [reply Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:102]: Left a TODO

File: Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean
Comment (MichaelStollBayreuth) [reply Mathlib/Analysis/Fourier/FiniteAbelian/PontryaginDuality.lean:59]: OK; I don't insist...

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
